text/0000-float-semantics.md

```diff
@@ -0,0 +1,287 @@
+`%` matches the behavior of `fmod` in C (this operation is not in the IEEE spec).
+When a floating-point value is just passed around (i.e., outside the operations above or any library-provided float operation), its representation bits do *not* change.
+
+Exceptions apply when the output of an operation is a NaN, and the operation is not a "bitwise" operation (unary `-`, `abs`, `copysign`).
+In that case, we generally follow [the same rules as LLVM](https://llvm.org/docs/LangRef.html#behavior-of-floating-point-nan-values).
+
+To be concrete, we first establish some terminology:
+A floating-point NaN value consists of a sign bit, a quiet/signaling bit, and a payload (which makes up the rest of the significand (i.e., the mantissa) except for the quiet/signaling bit). Rust assumes that the quiet/signaling bit being set to ``1`` indicates a quiet NaN (QNaN), and a value of ``0`` indicates a signaling NaN (SNaN). In the following we will hence just call it the "quiet bit".
+
+For the operations listed above, the following rules apply when a NaN value is returned:
+the result has a non-deterministic sign; the quiet bit and payload are non-deterministically chosen from the following set of options:
+
+- The quiet bit is set and the payload is all-zero. ("Preferred NaN" case)
+- The quiet bit is set and the payload is copied from any input operand that is a NaN.
+  ("Quieting NaN propagation" case) <br>
+  If the inputs and outputs do not have the same size (i.e., for `as` casts), then
+  - If the output is smaller than the input, low-order bits of the payload get dropped.
+  - If the output is larger than the input, the payload gets filled up with 0s in the low-order bits.
+- The quiet bit and payload are copied from any input operand that is a NaN.
+  ("Unchanged NaN propagation" case) <br>
+  If the inputs and outputs do not have the same size (i.e., for `as` casts), the same rules as for "quieting NaN propagation" apply, with one caveat: if the output is smaller than the input, droppig the low-order bits may result in a payload of 0; a payload of 0 is not possible with a signaling NaN (the all-0 significand encodes an infinity) so unchanged NaN propagation cannot occur with some inputs.
+- The quiet bit is set and the payload is picked from a target-specific set of
+  "extra" possible NaN payloads. The set can depend on the input operand values.
+  This set is empty on x86, ARM, and RISC-V (32bit and 64bit), but can be non-empty on other architectures. Targets where this set is non-empty should document this in a suitable location, e.g. their platform support page.
+  (For instance, on wasm, if any input NaN does not have the preferred all-zero   payload or any input NaN is an SNaN, then this set contains all possible payloads; otherwise, it is empty. On SPARC, this set consists of the all-one payload.)
```

Comment:
This section would be easier to understand if it separated the cases under "the quiet bit is set" and "the quiet bit is copied"
```suggestion
- The quiet bit is set, and
  - the payload is all-zero. ("Preferred NaN" case)
  - the payload is copied from any input operand that is a NaN. ("Quieting NaN propagation" case)  
  - the payload is picked from a target-specific set of "extra" possible NaN payloads. The set can depend on the input operand values. This set is empty on x86, ARM, and RISC-V (32bit and 64bit), but can be non-empty on other architectures. Targets where this set is non-empty should document this in a suitable location, e.g. their platform support page. (For instance, on wasm, if any input NaN does not have the preferred all-zero payload or any input NaN is an SNaN, then this set contains all possible payloads; otherwise, it is empty. On SPARC, this set consists of the all-one payload.)
- The quiet bit and payload are copied from any input operand that is a NaN. ("Unchanged NaN propagation" case)

When the payload is copied from an input with a different size (i.e., for `as` casts), then
- If the output is larger than the input, the payload gets filled up with zeros in the low-order bits.
- If the output is smaller than the input, low-order bits of the payload get dropped.

If the quiet bit and payload would be copied from a larger payload, dropping the low-order bits could result in a payload of 0; a payload of 0 is not possible with a signaling NaN (the all-0 significand encodes an infinity) so unchanged NaN propagation cannot occur with some inputs.
```

Reply:
I don't think this is better. It puts undue emphasis on the quiet bit.

Reply:
Okay, that's fair. Adding that emphasis was part of my intention, because not quieting an sNaN is the only case which diverges from the IEEE standard. In particular, the RFC seems to allows propagation of sNaN through "x * -1.0" without flipping the sign bit, whereas the sign of an sNaN should behave deterministically in all operations that don't quiet it.

Reply:
Yes we do deviate from IEEE deliberately here. But I assume few readers to be already familiar with what exactly IEEE says about SNaN so I didn't want to structure this as a diff to IEEE.

If x is an SNaN, then `x * -1.0` has arbitrary sign and is allowed to be
- the preferred QNaN
- a QNaN with the same payload as `x`
- an SNaN with the same payload as `x`

We also guarantee *more* than IEEE guarantees when it comes to the payload bits, so in the end it's incomparable.

Reply:
I'll leave this thread unresolved to see how others feel about presenting this as a flat list of 4 cases, vs structuring this around the quiet bit.

Reply:
Yes, and according to my understanding, when `x`  is an SNaN, IEEE requires
 - `x * -1.0` -> QNaN with unspecified sign and payload
 - `-x` -> SNaN with the payload of `x` and the opposite sign
 
(the latter is also required in this RFC)

 That is why I previously suggested the weaker (than IEEE) guarantee (see below) that would allow changing `x * -1.0` into `-x` (which LLVM currently does), while still preserving the property that the SNaN behaves like a signed value with unknown magnitude, until some operation depends on the magnitude and quiets it. I believe that would allow all practical optimizations while keeping to the spirit of the standard.

> If some part of a Rust floating point computation is equivalent to an operation
purely on the sign bit of a numeric input, then that same sign operation may be
used to perform the computation even when the input may be NaN.

Reply:
iirc wasm rules deterministic sign out, since basically every arithmetic operation is specified to return non-deterministic sign when returning a NaN

Reply:
Yeah, seems reasonable as a future option, and it is a subset from the current rfc.

With regards to LLVM's current optimizations; Alive2 does seem to implement the behavior as specified in the RFC (sign unspecified even in the snan-propagation case). I haven't managed to construct an example where LLVM would actually optimize like that, but probably it can.

>I think you want to make the sign deterministic when there is a signalling NaN input, but what is the deterministic spec for that sign? Given an arbitrary binary float operation, what should the sign of the returned NaN be?

To be clear, this only concerns the sign of any *signalling* NaNs.

(This actually got more optimization-friendly while formulating the response...)
Let `f(s,m)` be an operation with input and output in (sign,magnitude)-form (i.e. the sign bit `s` and the rest of the bits `m` separately). If a substitution `g(s,m)` is correct for all non-signalling inputs, and does not depend on the bit-representation of `m` (via `total_cmp`, `to_bits`,`is_nan`, etc.), the substitution could be made even if the input could be a signalling NaN. (For the record, this does not conform to the standard, but it's the closest that I've come up with while still allowing optimizations LLVM actually performs.)

So, why would this be reasonable? Because either the optimized code passes the SNaN to a different float-op that quiets it, or the SNaN magnitude gets returned as is, or the output doesn't depend on it at all. Some allowed transforms include:
    - `x * -1.0` -> `-x`
    - `x * 2.0` -> `x + x`
    - `powf(1.0, x)` -> `1.0`
    - `powf(x, 0.0)` -> `1.0`
    - `hypot(inf, x)` -> `inf`
    - `hypot(0, x)` -> `abs(x)`
(The first four of those are currently performed by LLVM)

> iirc wasm rules deterministic sign out, since basically every arithmetic operation is specified to return non-deterministic sign when returning a NaN

They do however [specify](https://webassembly.github.io/spec/core/exec/numerics.html#nan-propagation) that any returned NaN is "arithmetic", which is their definition for "the payload has its most significant bit set", which just so happens to be the quiet bit (but Wasm doesn't make the quiet/signalling distinction). So at least in this regard their rules follow the standard.

Reply:
@quaternic that characterizes the optimizations, but it's not what we'd need to adjust the semantics. We need a way to say what are all the possible behaviors of a Rust program after applying arbitrarily many of your and other optimizations.

Put differently: currently Miri carries a [pretty direct implementation](https://github.com/rust-lang/miri/blob/1bfe76b6bd4f533591897d32caf17f074b55b8a7/src/operator.rs#L97-L119) of the wording from this RFC, making sure we explore all possible options when a test is repeated sufficiently often. How would that change if we wanted to get your guarantee?

However, this is also getting somewhat off-topic for this RFC since we're not going to have this guarantee until someone convinces LLVM to provide it.  I also don't see the motivation for this change. Why is it useful to have a deterministic sign if the output is a signalling NaN? The quiet/signalling NaN distinction is largely useless anyway  given existing compiler transformations, and the sign of a NaN is still non-deterministic if the NaN is quiet (which is always a possibility), so... it seems like a completely pointless complication of our spec to me to try and make SNaN signs deterministic.

Reply:
Consider this program, which itself uses the default error handling, and calls a library that may utilize SNaNs as reserved operands
```rust
let mut data = somelib::init();
for x in &mut data {
    // negate
    *x = -(*x);
    // alternative: multiply by -1.0
    // *x = *x * -1.0;
}
somelib::eval(data);
```
As written (using `-x`), this program works correctly with this RFC: If there are signalling NaNs in the data initialized by the library, they are negated the same as any numeric values.

If the program instead uses the alternative expression `x * -1.0`, and the IEEE standard was followed, the library would receive quiet NaNs, and report an error (or return a quiet NaN to denote one).

However, the non-conforming specification proposed in the RFC allows the alternative program to instead change the signs in the data arbitrarily without quieting the SNaNs. The library would then *silently* produce incorrect results without either party realizing that an error should have occurred.

If allowing only the limited optimization scheme I described, the program would at least be guaranteed to behave as if it did a nondeterministic choice between  `-x` and `x * -1.0`, giving correct results or reporting errors, respectively, but never silently incorrect results.

Reply:
My claim, as above, is that the entire distinction of signalling and quiet NaNs is entirely pointless. IEEE had some plans here, and there is an alternative universe in which those planned panned out and they are useful. But in this universe, with today's compilers, they are not. You can't meaningfully use them as reserved values as one can freely create them with `from_bits`. You can't check whether they flow into arithmetic since you can't observe the status bits. You can't check if there was an arithmetic operation since SNaNs might or might not be quieted. So it's not worth trying to support them in any meaningful way, other than making sure we don't generate them "out of thin air".

Anyone who begs to differ is free to write up an RFC detailing documentation and specification; I don't consider that a good use of my time.

Reply:
I could continue with more opinions on your other points, but it indeed isn't very relevant here. We certainly agree that this RFC can't make more guarantees on this unless LLVM does.